android/src/main/java/com/twiliorn/library/PatchedVideoView.java

```diff
@@ -45,6 +64,108 @@ public void run() {
             });
         }
         super.onFrame(frame);
+
+        if (captureThisFrame) {
+            captureThisFrame = false;
+            saveVideoFrame(frame);
+        }
+    }
+
+    /**
+     * Create an NV21Buffer with the same pixel content as the given I420 buffer. Ripped from here: https://stackoverflow.com/a/55158228
+     */
+    private byte[] createNV21Data(VideoFrame.I420Buffer i420Buffer) {
+        final int width = i420Buffer.getWidth();
+        final int height = i420Buffer.getHeight();
+        final int chromaStride = width;
+        final int chromaWidth = (width + 1) / 2;
+        final int chromaHeight = (height + 1) / 2;
+        final int ySize = width * height;
+        final ByteBuffer nv21Buffer = ByteBuffer.allocateDirect(ySize + chromaStride * chromaHeight);
+        // We don't care what the array offset is since we only want an array that is direct.
+        @SuppressWarnings("ByteBufferBackingArray") final byte[] nv21Data = nv21Buffer.array();
+        for (int y = 0; y < height; ++y) {
+            for (int x = 0; x < width; ++x) {
+                final byte yValue = i420Buffer.getDataY().get(y * i420Buffer.getStrideY() + x);
+                nv21Data[y * width + x] = yValue;
+            }
+        }
+        for (int y = 0; y < chromaHeight; ++y) {
+            for (int x = 0; x < chromaWidth; ++x) {
+                final byte uValue = i420Buffer.getDataU().get(y * i420Buffer.getStrideU() + x);
+                final byte vValue = i420Buffer.getDataV().get(y * i420Buffer.getStrideV() + x);
+                nv21Data[ySize + y * chromaStride + 2 * x + 0] = vValue;
+                nv21Data[ySize + y * chromaStride + 2 * x + 1] = uValue;
+            }
+        }
+        return nv21Data;
+    }
+
+    private void sendEvent(String event, WritableMap params) {
+        ((ReactContext) this.getContext()).getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
+                .emit(event, params);
+    }
+
+    private void saveVideoFrame(VideoFrame frame) {
+        frame.retain();
```

Comment:
🔧 You don't need to `retain`/`release` the frame unless you need to read the contents of `frame` after `saveVideoFrame()` returns. In your current code, you're reading the contents before the method returns.